packages/multichain/src/middlewares/MultichainMiddlewareManager.ts

```diff
@@ -0,0 +1,132 @@
+import type {
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import type {
+  Json,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+
+import type { ExternalScopeString } from '../scope/types';
+
+export type ExtendedJsonRpcMiddleware = {
+  (
+    req: JsonRpcRequest & { scope: string },
+    res: PendingJsonRpcResponse<Json>,
+    next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+  ): void;
+  destroy?: () => void | Promise<void>;
+};
+
+type MiddlewareKey = {
+  scope: ExternalScopeString;
+  origin: string;
+  tabId?: number;
+};
+type MiddlewareEntry = MiddlewareKey & {
+  middleware: ExtendedJsonRpcMiddleware;
+};
+
+export class MultichainMiddlewareManager {
+  #middlewares: MiddlewareEntry[] = [];
+
+  #getMiddlewareEntry({
+    scope,
+    origin,
+    tabId,
+  }: MiddlewareKey): MiddlewareEntry | undefined {
+    return this.#middlewares.find((middlewareEntry) => {
+      return (
+        middlewareEntry.scope === scope &&
+        middlewareEntry.origin === origin &&
+        middlewareEntry.tabId === tabId
+      );
+    });
+  }
+
+  #removeMiddlewareEntry({ scope, origin, tabId }: MiddlewareEntry) {
+    this.#middlewares = this.#middlewares.filter((middlewareEntry) => {
+      return (
+        middlewareEntry.scope !== scope ||
+        middlewareEntry.origin !== origin ||
+        middlewareEntry.tabId !== tabId
+      );
+    });
+  }
+
+  addMiddleware(middlewareEntry: MiddlewareEntry) {
+    const { scope, origin, tabId } = middlewareEntry;
+    if (!this.#getMiddlewareEntry({ scope, origin, tabId })) {
+      this.#middlewares.push(middlewareEntry);
+    }
+  }
+
+  #removeMiddleware(middlewareEntry: MiddlewareEntry) {
+    // When the destroy function on the middleware is async,
+    // we don't need to wait for it complete
+    // eslint-disable-next-line no-void
+    void middlewareEntry.middleware.destroy?.();
```

Comment:
I don't know why ESLint recommends using `void`, but it seems like it wouldn't prevent an uncaught promise error, which I think is what we really want to avoid here.

What about:

``` suggestion
    Promise.resolve(middlewareEntry.middleware.destroy?.()).catch(() => {
      // do nothing
    });
```